### ISSUE TITLE
auStack before every call contain return address

### ISSUE THREAD

**Issue by bravosierra99** [NationalSecurityAgency/ghidra#3414]
Trying to determine the root cause of this.  I some functions (not all) every call instruction is proceeded by something like the following.

```    auStack1360[0] = 0x401b07;
    sVar7 = wcslen(&local_108);```

`0x401b07` is the address of the instruction following the call.  I thought maybe this had something to do with the function signature... but I don't see anything wrong or weird.  I tried playing with the options and still failed miserably.

Other functions in this program all decompile just fine... and then some of them have this for every single call in the function

the signature for one of the bad functions is
`DWORD fastcall func (wchar_t * param_1, LPCWSTR param_2, undefined8 * key_stuff)`  everything else in the function Signature is standard

**Comment by Danil6969** [comment]
It's return address following after call. If stack is complicated enough it's not optimized out as stack propagation algorithm doesn't catch all the cases where stack variable is unused.

**Comment by quietust** [comment]
I see this happen all the time in 32-bit x86 code when a structure is dynamically allocated on the stack to be passed by value to a function (e.g. if a function was accidentally coded as `void doStuff(std::string val)` instead of `void doStuff(std::string &val)`), but only when I try to actually insert an instance of the structure into the stack frame.

**Comment by bravosierra99** [comment]
@quietust  So this is 64bit code and this happened with the vanilla analysis settings. 

@Danil6969 right I knew what the number was just not why it was there. I wouldn't really call this that complicated of a function, though it does have 40-50 calls in it. Is there anything you can do to help the decompile deal with it?

**Comment by astrelsky** [comment]
Is there an alloca or something of that nature in the function? If so its probably that part of the pcode injection isn't being considered for optimization or something. Just a guess though.

**Comment by bravosierra99** [comment]
It's from the CSAW CTF, here is the content.  There are some malloc calls.
https://pastebin.com/ENtLMK1Z

**Comment by bravosierra99** [comment]
my initial thoughts are stack misalignment of some sort.  Probably due to function signature, but they just look normal and standard?

**Comment by astrelsky** [comment]
> It's from the CSAW CTF, here is the content.  There are some malloc calls.
> https://pastebin.com/ENtLMK1Z
> 

I did mean `alloca` as this shifts the stack pointer. The actual function you may be seeing is `alloca_probe` or `chk_stack`. In the decompiler window above the function signature is there a comment about pcode injection?

**Comment by bravosierra99** [comment]
nope, just this 

`/* WARNING: Could not reconcile some variable overlaps */`

**Comment by bravosierra99** [comment]
it's from a public source CTF, I can provide the binary as well.

**Comment by bravosierra99** [comment]
I just reloaded it, and that Warning is not actually there, so must have been something from me messing around.  All the auStack stuff is still there obviously.

**Comment by astrelsky** [comment]
> nope, just this 
> 
> `/* WARNING: Could not reconcile some variable overlaps */`
> 

Ok. I see a `___chkstk_ms` function call. You might want to take a close look at it's behavior. If it modifies the stack pointer then you should try creating a call other override to inject pcode where the call is to correct it. You might get away with applying the injection for `chk_stack` to it. You can do this in the drop down box at the bottom left corner of the edit function window for `___chkstk_ms`.

**Comment by bravosierra99** [comment]
so I don't have that option in the edit window, but I do have alloca_probe which is I understand the same function.  I tried that one (and actually the other I have guard_dispatch_icall).  Neither seemed to change much other than the specific decompilation around the __chkstk_ms call.   hmmmmm.

when I look at the pcode I see the exact instruction that stores the RIP, but then again I see that also in functions where I'm not getting this erroneous auStack stuff.

**Comment by bravosierra99** [comment]
I'll keep looking at it. You mentioned some other stuff that I'll have to figure out how to try. Thanks.

**Comment by bravosierra99** [comment]
so this is interesting, if I start nopping out the chkstk_ms call.... it breaks most of the function (only displays the top quarter?) but the part that remains is FIXED.  when I finish nopping it out (all 5 bytes of the call opcode...) the missing parts of the function come back and it's broken again...

I did look inside the function, and it subs 8 3x from RSP and then only adds back 8 2x.  But then there is the RET instruction, does that automagically count as adding 8 back to RSP (in pcode)?

**Comment by bravosierra99** [comment]
nah, thats just consuming the RIP off the stack, so that actually looks correct

**Comment by bravosierra99** [comment]
so while I can't figure out how or why... the two functions that are messed up are the two  functions that call this `__chkstk_ms`.  So I believe you are right... but when I look at that function it seems perfectly correct.  Outside the call the RIP is pushed.  Inside two regs are pushed.  Bunch of stuff happens.  Two regs are popped.  RIP is popped.  Return RIP. that seems like perfect stack alignment.  Nopping out the call to it doesn't fix anything, even though that should ensure the stack is aligned if it's causing the problem...

**Comment by bravosierra99** [comment]
hmmm
https://github.com/NationalSecurityAgency/ghidra/issues/1888

**Comment by bravosierra99** [comment]
GOT IT, ok so AFTER the `__chkstk_ms ` the compiler had injected an `sub RSP, RAX`.  I'm guessing inlining something related to the `__chkstk_ms` call.  Anyways, that was making the stack super confused and weird.  I nopped that instruction out... and suddenly everything in the decompiler was perfect.
